description: Add flag to disable creation of WebRTC RTCPeerConnection object
author: Wengling Chen <feiyu2817@gmail.com>

---
 chrome/browser/ungoogled_flag_entries.h                           | 4 ++++
 third_party/blink/common/features.cc                              | 3 +++
 third_party/blink/public/common/features.h                        | 3 +++
 .../blink/renderer/modules/peerconnection/rtc_peer_connection.cc  | 8 ++++++++
 4 files changed, 18 insertions(+)

--- a/chrome/browser/ungoogled_flag_entries.h
+++ b/chrome/browser/ungoogled_flag_entries.h
@@ -4,6 +4,10 @@
 
 #ifndef CHROME_BROWSER_UNGOOGLED_FLAG_ENTRIES_H_
 #define CHROME_BROWSER_UNGOOGLED_FLAG_ENTRIES_H_
+    {"disable-peerconnection",
+     "Disable WebRTC peerconnection",
+     "Prevent JavaScript from creating RTCPeerConnection object. ungoogled-chromium-android flag.",
+     kOsAndroid, FEATURE_VALUE_TYPE(blink::features::kDisablePeerConnection)},
     {"set-ipv6-probe-false",
      "SetIpv6ProbeFalse",
      "Forces the result of the browser's IPv6 probing (i.e. IPv6 connectivity test) to be unsuccessful. This causes IPv4 addresses to be prioritized over IPv6 addresses. Without this flag, the probing result is set to be successful, which causes IPv6 to be used over IPv4 when possible. ungoogled-chromium flag.",
--- a/third_party/blink/common/features.cc
+++ b/third_party/blink/common/features.cc
@@ -579,6 +579,9 @@ BASE_FEATURE(kDevToolsImprovedNetworkErr
 
 BASE_FEATURE(kDevToolsAdsPanel, base::FEATURE_ENABLED_BY_DEFAULT);
 
+// Disable PeerConnection creations. ungoogled-chromium-android flag.
+BASE_FEATURE(kDisablePeerConnection, base::FEATURE_ENABLED_BY_DEFAULT);
+
 BASE_FEATURE(kDisableArrayBufferSizeLimitsForTesting,
              base::FEATURE_DISABLED_BY_DEFAULT);
 
--- a/third_party/blink/public/common/features.h
+++ b/third_party/blink/public/common/features.h
@@ -405,6 +405,9 @@ BLINK_COMMON_EXPORT BASE_DECLARE_FEATURE
 // Enables the DevTools 'Application > Application > Ads' panel.
 BLINK_COMMON_EXPORT BASE_DECLARE_FEATURE(kDevToolsAdsPanel);
 
+// Disable PeerConnection creations. ungoogled-chromium-android flag.
+BLINK_COMMON_EXPORT BASE_DECLARE_FEATURE(kDisablePeerConnection);
+
 // TODO(https://crbug.com/1201109): temporary flag to disable new ArrayBuffer
 // size limits, so that tests can be written against code receiving these
 // buffers. Remove when the bindings code instituting these limits is removed.
--- a/third_party/blink/renderer/modules/peerconnection/rtc_peer_connection.cc
+++ b/third_party/blink/renderer/modules/peerconnection/rtc_peer_connection.cc
@@ -638,6 +638,14 @@ RTCPeerConnection* RTCPeerConnection::Cr
     ExecutionContext* context,
     const RTCConfiguration* rtc_configuration,
     ExceptionState& exception_state) {
+  // TODO: per-site UI
+  if (base::FeatureList::IsEnabled(features::kDisablePeerConnection)) {
+    exception_state.ThrowDOMException(
+        DOMExceptionCode::kNotSupportedError,
+        "RTCPeerConnection creations are prohibited in this browser.");
+    return nullptr;
+  }
+
   if (context->IsContextDestroyed()) {
     exception_state.ThrowDOMException(
         DOMExceptionCode::kNotSupportedError,
